src/shims/foreign_items.rs

```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     /// Emulates calling a foreign item using its name.
     fn emulate_foreign_item_by_name(
         &mut self,
+        def_id: DefId,
         link_name: Symbol,
         abi: Abi,
         args: &[OpTy<'tcx, Tag>],
         dest: &PlaceTy<'tcx, Tag>,
     ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
         let this = self.eval_context_mut();
 
+        let tcx = this.tcx.tcx;
+
+        // First deal with any external C functions in linked .so file
+        // (if any SO file is specified).
+        if this.machine.external_c_so_lib.as_ref().is_some() {
+            if let Some(local_id) = def_id.as_local() {
+                if let Some(extern_c_fct_rep) = ExternalCFuncDeclRep::from_hir_node(
+                    &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)),
+                    link_name,
+                ) {
```

Comment:
More specifically, what you can use instead of the HIR is `tcx.fn_sig(def_id)` (which also has `input` and `output` fields, but the types are `Ty`, which you can take apart for the individual types by matching on the result of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.kind